test/test-ratelim.c

```diff
@@ -208,6 +214,12 @@ echo_listenercb(struct evconnlistener *listener, evutil_socket_t newsock,
 		bufferevent_add_to_rate_limit_group(bev, ratelim_group);
 	++n_echo_conns_open;
 	bufferevent_enable(bev, EV_READ|EV_WRITE);
+	return;
+err:
+	if (bev)
+		bufferevent_free(bev);
+	fprintf(stderr, "Couldn't create bufferevent\n");
```

Comment:
I don't really understand how libevent works, so there are a few things I'm not quite sure (mainly about when a socket needs to be closed). I need to add:
```
 if (newsock != EVUTIL_INVALID_SOCKET)
		EVUTIL_CLOSESOCKET(newsock);
```
under err, right?
And I need to convert **every** assert to goto err?

Reply:
>I don't really understand how libevent works, so there are a few things I'm not quite sure (mainly about when a socket needs to be closed)

listen callback of event listener need to handle fd somehow, usually it creates bufferevent from it for later use, however, if this fails, you need to close fd manually so that it won't leak (in case of success bufferevent will do this either with `BEV_OPT_CLOSE_ON_FREE` or manually in errorcb)

By the way, there is a short intro into libevent, that you can find here - https://libevent.org/libevent-book/

> I need to add:

No, just `evutil_closesocket()` (BTW it is enough to use `evutil_closesocket` instead of a macro `EVUTIL_CLOSESOCKET`)

>And I need to convert every assert to goto err?

Yes